cloudwatch_logs_common/src/log_publisher.cpp

```diff
@@ -108,102 +107,153 @@ Aws::CloudWatchLogs::ROSCloudWatchLogsErrors LogPublisher::StopPublisherThread()
-              "Unable to send logs to CloudWatch and retried, dropping this batch of logs ...");
+            AWS_LOG_WARN(__func__, "Unable to send logs to CloudWatch, retrying ...");
+            Aws::CloudWatchLogs::ROSCloudWatchLogsErrors get_token_status =
+              this->cloudwatch_facade_->GetLogStreamToken(this->log_group_, this->log_stream_,
+                                                          next_token);
+            if (CW_LOGS_SUCCEEDED != get_token_status) {
+              AWS_LOG_WARN(__func__,
+                           "Unable to obtain the sequence token to use, quit attempting to send "
+                           "logs to CloudWatch");
+              break;
+            }
           }
+          tries--;
+        }
+        if (CW_LOGS_SUCCEEDED != send_logs_status) {
+          AWS_LOG_WARN(
+            __func__,
+            "Unable to send logs to CloudWatch and retried, dropping this batch of logs ...");
         }
       }
-      // TODO: For now we're just going to discard logs that fail to get sent. Later we may add some
-      // retry strategy
-
-      /* Null out the class reference before unlocking the object. The external thread will be
-         looking to key off if the shared object is locked or not to know it can start again, but
-         the PublishLogs function checks if the the pointer is null or not. */
-      this->shared_logs_.store(nullptr, std::memory_order_release);
-      shared_logs_obj->freeDataAndUnlock();
     }
-    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    // TODO: For now we're just going to discard logs that fail to get sent. Later we may add some
+    // retry strategy
+
+    /* Null out the class reference before unlocking the object. The external thread will be
+       looking to key off if the shared object is locked or not to know it can start again, but
+       the PublishLogs function checks if the the pointer is null or not. */
+    this->shared_logs_.store(nullptr, std::memory_order_release);
+    shared_logs_obj->freeDataAndUnlock();
+  }
+}
+
+void LogPublisher::Run()
+{
+  Aws::String next_token;
+  this->run_state_ = LOG_PUBLISHER_RUN_CREATE_GROUP;
+  
+  while (this->thread_keep_running_.load(std::memory_order_acquire)) {
+    LogPublisherRunState previous_state = this->run_state_;
+    switch (this->run_state_) {
+      case LOG_PUBLISHER_RUN_CREATE_GROUP:
+        CreateGroup();
+        break;
+      case LOG_PUBLISHER_RUN_CREATE_STREAM:
+        CreateStream();
+        break;
+      case LOG_PUBLISHER_RUN_INIT_TOKEN:
+        InitToken(next_token);
+        break;
+      case LOG_PUBLISHER_RUN_SEND_LOGS:
+        SendLogs(next_token);
+        break;
+      default:
+        AWS_LOGSTREAM_ERROR(__func__, "Unknown state!");
+        break;
+    }
+    if (previous_state == this->run_state_) {
+      // Don't sleep between state changes for a faster startup
+      std::this_thread::sleep_for(std::chrono::milliseconds(1000)); 
```

Comment:
Note: see my comment about the thread-safe containers. What you want is actually what is described in C++ concurrency in action, Chapter 6, Listing 6.9. Another thread pushes in a FIFO object (of size 1 in this case) and we would pull in this thread using `wait_and_pop`. We would only need to add a `shutdown` method to this container which would move it to a "lame duck" mode where all waiting methods would immediately return with an error (e.g. failed to pop). This would allow us to exit this thread when the application finishes.

Reply:
I agree that the condition_variable could be used instead, but that's a bigger change than I want to make in this PR. This sleep is used as a small backoff between activity as the loop also takes care of retries for all the create states. In the happy case this shouldn't slow us down unless someone is trying to increase the publish rate below 1 second, which isn't advisable for cost reasons, and in the not so happy case it provides a slight back off to prevent throttling exceptions.